### PR TITLE
Move around packaging, refactor out process, simplify message feedback

### PR DESCRIPTION
**Pull request by loosebazooka**
This CL is really just moving a bunch of code around, removes the "internal" package and goes for package private to hide most implementation details. #472 

Adds a CommandExecutor that's derived from @coollog's work in minikube tooling.

Feedback now is just through a single line listener.

### CODE REVIEW

File: src/main/java/com/google/cloud/tools/managedcloudsdk/process/CommandExecutor.java [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    return this;
+  }
+
+  /** Sets the environment variables to run the command with. */
+  public CommandExecutor setEnvironment(Map<String, String> environmentMap) {
+    this.environment = environmentMap;
+    return this;
+  }
+
+  // @VisibleForTesting
```
Comment (elharo) [inline src/main/java/com/google/cloud/tools/managedcloudsdk/process/CommandExecutor.java]: Is it or isn't it? I.e. why the comment?

Comment (loosebazooka) [reply src/main/java/com/google/cloud/tools/managedcloudsdk/process/CommandExecutor.java]: oops I'll fix all those.

File: src/main/java/com/google/cloud/tools/managedcloudsdk/install/Downloader.java [hunk 2]
```diff
@@ -14,60 +14,55 @@
...
-      throw new NotDirectoryException(
-          "Cannot download to " + destinationDir + " because it is not a directory");
-    }
-    if (Files.exists(destinationFile)) {
-      throw new FileAlreadyExistsException(
-          "Cannot write to " + destinationFile + " because it already exists");
-    }
-    if (!Files.exists(destinationDir)) {
-      Files.createDirectories(destinationDir);
+    if (!Files.exists(destinationFile.getParent())) {
```
[the remark this answers is not in the document]
Comment (loosebazooka) [reply src/main/java/com/google/cloud/tools/managedcloudsdk/install/Downloader.java]: Yeah, I didn't mean for it to be part of this CL, but basically the installer-glue is going to ensure the environment.

File: src/main/java/com/google/cloud/tools/managedcloudsdk/install/Extractor.java [hunk 3]
```diff
@@ -14,37 +14,40 @@
...
+    this.messageListener = messageListener;
   }
 
+  /** Extract and return a {@link Path} to the Cloud SDK home directory. */
   @Override
   public Path call() throws IOException {
-    if (listener != null) {
-      listener.message("Extracting archive: " + archive.toString());
-    }
+    messageListener.message("Extracting archive: " + archive.toString());
```
Comment (coollog) [inline src/main/java/com/google/cloud/tools/managedcloudsdk/install/Extractor.java]: `.toString()` here and below are unnecessary as well.

File: src/test/java/com/google/cloud/tools/managedcloudsdk/install/ExtractorTest.java [hunk 4]
```diff
@@ -49,16 +57,16 @@ public Void answer(InvocationOnMock invocationOnMock) throws Throwable {
...
     Extractor extractor =
-        new ConfigurableExtractor<>(
-            extractionSource, extractionDestination, mockProvider, mockListener);
+        new Extractor<>(
+            extractionSource, extractionDestination, mockExtractorProvider, mockMessageListener);
     Path cloudSdkHomeUnderTest = extractor.call();
 
     Assert.assertEquals(expectedCloudSdkHome, cloudSdkHomeUnderTest);
-    Mockito.verify(mockListener, Mockito.times(1))
+    Mockito.verify(mockMessageListener, Mockito.times(1))
```
Comment (coollog) [inline src/test/java/com/google/cloud/tools/managedcloudsdk/install/ExtractorTest.java]: `Mockito.times(1)` is used by default.

File: src/test/java/com/google/cloud/tools/managedcloudsdk/install/ExtractorTest.java [hunk 5]
```diff
@@ -49,16 +57,16 @@ public Void answer(InvocationOnMock invocationOnMock) throws Throwable {
...
         .message("Extracting archive: " + extractionSource.toString());
```
Comment (coollog) [inline src/test/java/com/google/cloud/tools/managedcloudsdk/install/ExtractorTest.java]: `toString()` here as well.

File: src/test/java/com/google/cloud/tools/managedcloudsdk/install/InstallerFactoryTest.java [hunk 6]
```diff
@@ -45,24 +43,17 @@
...
         expectedInstallScriptProviderClass,
-        ((LatestInstaller) installer).getInstallScriptProvider().getClass());
-  }
-
-  @Test
-  public void testNewInstaller_fixedVersion() throws BadCloudSdkVersionException {
-    Installer installer =
-        new InstallerFactory(new Version("1.0.0"), OS, false).newInstaller(null, null);
-    Assert.assertEquals(NoOpInstaller.class, installer.getClass());
+        ((Installer) installer).getInstallScriptProvider().getClass());
```
Comment (coollog) [inline src/test/java/com/google/cloud/tools/managedcloudsdk/install/InstallerFactoryTest.java]: Redundant cast to Installer

File: src/test/java/com/google/cloud/tools/managedcloudsdk/install/InstallerFactoryTest.java [hunk 7]
```diff
@@ -45,24 +43,17 @@
...
 
   @Parameterized.Parameter(1)
   public Class<? extends InstallScriptProvider> expectedInstallScriptProviderClass;
 
   @Test
   public void testNewInstaller_latestVersion() {
-    Installer installer = new InstallerFactory(Version.LATEST, OS, false).newInstaller(null, null);
-    Assert.assertEquals(LatestInstaller.class, installer.getClass());
+    Installer installer = new InstallerFactory(os, false).newInstaller(null, null);
+    Assert.assertEquals(Installer.class, installer.getClass());
```
Comment (coollog) [inline src/test/java/com/google/cloud/tools/managedcloudsdk/install/InstallerFactoryTest.java]: Now that there is only one Installer type, the class check is unnecessary.

File: src/test/java/com/google/cloud/tools/managedcloudsdk/install/UnixInstallScriptProviderTest.java [hunk 8]
```diff
@@ -14,13 +14,13 @@
...
 
-package com.google.cloud.tools.managedcloudsdk.internal.install;
+package com.google.cloud.tools.managedcloudsdk.install;
 
 import java.util.Arrays;
 import org.junit.Assert;
 import org.junit.Test;
 
-/** Tests for {@link UnixInstallScriptProvider} */
+/** Tests for {@link com.google.cloud.tools.managedcloudsdk.install.UnixInstallScriptProvider} */
```
Comment (coollog) [inline src/test/java/com/google/cloud/tools/managedcloudsdk/install/UnixInstallScriptProviderTest.java]: Is this full package path intentional?

File: src/test/java/com/google/cloud/tools/managedcloudsdk/install/WindowsInstallScriptProviderTest.java [hunk 9]
```diff
@@ -14,13 +14,13 @@
...
 
-package com.google.cloud.tools.managedcloudsdk.internal.install;
+package com.google.cloud.tools.managedcloudsdk.install;
 
 import java.util.Arrays;
 import org.junit.Assert;
 import org.junit.Test;
 
-/** Tests for {@link WindowsInstallScriptProvider} */
+/** Tests for {@link com.google.cloud.tools.managedcloudsdk.install.WindowsInstallScriptProvider} */
```
Comment (coollog) [inline src/test/java/com/google/cloud/tools/managedcloudsdk/install/WindowsInstallScriptProviderTest.java]: Same here.

File: src/test/java/com/google/cloud/tools/managedcloudsdk/install/InstallerFactoryTest.java (hunk 7 above)
Comment (loosebazooka) [reply src/test/java/com/google/cloud/tools/managedcloudsdk/install/InstallerFactoryTest.java]: hah, I'm dumb.

**Pull request merged**